#6588: Add output tensors support for matmul

KalaivaniMCW opened the pull request:
- made changes to remove std::reference_wrapper in moreh_matmul and moreh_matmul_backward
- Primary issue #5044 
- CI test

KalaivaniMCW committed:
#6588: Add output tensors support for matmul

eyonland reviewed (commented):
Please add a test showing that the optional output tensor does not even need to be provided.  And one that shows that the optional output tensor was provided and returned back by the op.

eyonland reviewed (changes requested):
Please add a test showing that the optional output tensor does not even need to be provided.  And one that shows that the optional output tensor was provided and returned back by the op.

KalaivaniMCW commented:
> Please add a test showing that the optional output tensor does not even need to be provided. And one that shows that the optional output tensor was provided and returned back by the op.

Hi @eyonland , 
I have made changed in moreh_matmul_backward to receive and return the optional output_tensor. I have also added the test case with optional_output. 
> In case of moreh_matmul_backward, since the returned output is a vector of Tensors, and I have made changes so that optional output_tensor is stored at the 2nd index of the result. Is this what you expect ?

> In case of moreh_matmul, I have a confusion since the structure of moreh_matmul returns only one Tensor and 
`std::optional<const Tensor> output_tensor ` part was already present in the implementation - should I be adding another optional output_tensor or is this it ? The output_tensor value is returned as the result at line 174 `return output_tensor.value(); ` 
> I have added a test case with optional output being passed. If any additional tensor is to be returned then the return type of moreh_matmul must be changed to handle vector of Tensors `std::vector<Tensor>`

Existing tests still work even without passing optional output tensor, incase of both forward and backward.

KalaivaniMCW commented:
> 
> Hi @eyonland , I have made changed in moreh_matmul_backward to receive and return the optional output_tensor. I have also added the test case with optional_output.
> 
> * In case of moreh_matmul_backward, since the returned output is a vector of Tensors, and I have made changes so that optional output_tensor is stored at the 2nd index of the result. Is this what you expect ?
> * I have added a test case : test_moreh_matmul_backward_with_optional_output_tensor where optional output tensor is passed to moreh_matmul_backward and the same is received in result
> * In case of moreh_matmul, I have a confusion since the structure of moreh_matmul returns only one Tensor and
>   `std::optional<const Tensor> output_tensor ` part was already present in the implementation - should I be adding another optional output_tensor or is this it ? The output_tensor value is returned as the result at line 174 `return output_tensor.value(); `
> * I have added a test case with optional output being passed. If any additional tensor is to be returned then the return type of moreh_matmul must be changed to handle vector of Tensors `std::vector<Tensor>`
> 
> Existing tests still work even without passing optional output tensor, incase of both forward and backward.

Hi @razorback3 and @seunghwan100, cc: @eyonland, 
> In existing (main) moreh_matmul (forward),  an optional output_tensor std::optional<const Tensor> output_tensor   is already present which stores the output of the matmul operation and is returned as a single Tensor return output_tensor.value(); 
Is this the expected changes requested in issue [#5044](https://github.com/tenstorrent-metal/tt-metal/issues/5044) or should I be adding another tensor that is to be used for some other purpose?
This is the same doubt we have for few other moreh ops (like moreh arange, sum, softmax) as well.
> For moreh_matmul (backward), I have added std::optional<const Tensor> output_tensor and returned it along with the vector of outputs from the moreh_matmul_backward operation.
We wanted to confirm once before merging these changes

seunghwan100 commented:
- First of all, the implementation of this PR is wrong.
I think you need to understand what are_required_outputs arg of moreh_groupnorm_backward means.
And you also need to understand how it is used.
That could help you understand.
This can be applied to moreh_matmul_backward as well.
test_moreh_groupnorm.py also could help you understand.

https://github.com/tenstorrent-metal/tt-metal/blob/07a48b7836cf1c0127bfe36c8426302f6e354391/tt_eager/tt_dnn/op_library/moreh_groupnorm_backward/moreh_groupnorm_backward_op.hpp#L124

https://github.com/tenstorrent-metal/tt-metal/blob/07a48b7836cf1c0127bfe36c8426302f6e354391/tests/tt_eager/python_api_testing/unit_testing/misc/test_moreh_groupnorm.py#L386

- It seems that the meaning of the word "optional output sensor" is misunderstood to you.
Again, please look carefully at the implementation of moreh_groupnorm_backward

- moreh_matmul do not need a are_required_outputs arg because the number of moreh_matmul's output is 1.

seunghwan100 commented on tt_eager/tt_dnn/op_library/moreh_matmul_backward/moreh_matmul_backward_op.cpp:
```diff
@@ -30,8 +30,9 @@ inline bool is_dot_backward(const Tensor& output_grad, const Tensor& input, cons
-    std::optional<std::reference_wrapper<const Tensor>> other_grad,
+    std::optional<const Tensor> input_grad,
+    std::optional<const Tensor> other_grad,
+    std::optional<const Tensor> output_tensor,
```
This is unnecessary

seunghwan100 replied:
This is unnecessary.

seunghwan100 commented:
@arakhmati @eyonland 
I heard that you guys provide provide the API with more polished implementation.
https://github.com/tenstorrent-metal/tt-metal/issues/5163#issuecomment-2000424686

That API could be applied to moreh_matmul_backward.

eyonland commented:
Hey @seunghwan100 , @arakhmati and I can't get to updating the API anytime soon.  Can we get your guidance on how MCW can best help with getting the code updated?  

There seems to be two different objectives for optional output tensors.  If you agree, can we find a path to achieve both these goals using the existing API?

1. We want to give the user the ability to allocate the output tensor before the operation is called if desired.
2. In the backward pass there are times when we do not want the output tensors to be updated.  The clear example is layer_norm and group_norm where the variables like `are_needed_outputs` and `std::vector<bool> &are_required_outputs` are used to let the op know what output tensors should receive updates.

I'm not seeing that we need the second objective on the matmul or matmul_backward.   

- Would you mind adding an additional commit to this branch `kalaivani/output_tensors_matmul` so that we can all clearly see what the proper change should be for this effort?   
- Would you be willing to be a code reviewer on this effort going forward to help us make sure all of us are moving forward in agreement?

This will help me and MCW understand the requirements that y'all need.  My hope is that we can help you deliver on all these ops much faster.   Thank you very much for your help.

seunghwan100 commented:
> I'm not seeing that we need the second objective on the matmul or matmul_backward.

The second objective is needed on the matmul_backward because input_grad or other_grad can be null.
I think it's better for me to create a new branch, work on it, and then submit a PR.
Is it okay?

>  @arakhmati and I can't get to updating the API anytime soon.

I would like to know the expected timeline.

seunghwan100 commented:
https://github.com/tenstorrent-metal/tt-metal/issues/5044 will resolve [1. We want to give the user the ability to allocate the output tensor before the operation is called if desired.]

https://github.com/tenstorrent-metal/tt-metal/issues/5163 will resolve [2. In the backward pass there are times when we do not want the output tensors to be updated. The clear example is layer_norm and group_norm where the variables like are_needed_outputs and std::vector<bool> &are_required_outputs are used to let the op know what output tensors should receive updates.]

Yes, I agree that it's confusing when two objectives are mixed together.

eyonland commented:
Yes, creating a new branch would be best.

seunghwan100 commented:
> Yes, creating a new branch would be best.

I will done the work within this week.

seunghwan100 commented:
@KalaivaniMCW @eyonland 
I made a PR https://github.com/tenstorrent-metal/tt-metal/pull/6934

KalaivaniMCW commented:
Closing this PR according to https://github.com/tenstorrent-metal/tt-metal/issues/5044#issuecomment-2032963952
